### PR TITLE
Cross-dataset: EMA refinement — push TF/AF new bests further

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2899 (merged) established two new bests with corrected EMA decay=0.999:
- **TandemFoil:** 26.134 (was 30.10 — 13.2% improvement) at ep123, still improving
- **AirfRANS:** 0.000727 (was 0.001236 — 41.2% improvement) at ep206

Now explore what lies below these new bests:

1. **TandemFoil**: EMA is now confirmed essential. The TF run was still descending at ep123 (timeout). The next step is to compound EMA with the LR descent that was working before. lr=1e-4 + EMA at 3L/192d is untested. Also: does gc=0.5 compound with EMA on TF?

2. **AirfRANS**: The 0.000727 result was at T_max=5. Does T_max=10 (the intermediate scheduler) push further? Also: is the improvement architecture-agnostic (does it hold at different depths)?

## Instructions

**NOTE: No --no-use-ema flag. Use --ema-decay 0.999 (default is now 0.9999, so specify explicitly).**

Run 1 — TandemFoil lr=1e-4 + EMA decay=0.999 (lower LR with EMA):
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion \
  --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 \
  --wandb-group robin-ema-refinement
```

Run 2 — TandemFoil lr=1.25e-4 + gc=0.5 + EMA decay=0.999 (softer clip with EMA):
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion \
  --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 \
  --wandb-group robin-ema-refinement
```

Run 3 — AirfRANS T_max=10 + EMA decay=0.999 (intermediate scheduler with EMA):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --wandb-group robin-ema-refinement
```

Run 4 — AirfRANS lr=7e-4 + T_max=5 + EMA decay=0.999 (reproduce and extend new best at seed=43):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --seed 43 \
  --wandb-group robin-ema-refinement
```

Report `val_primary/surface_pressure_mae` for TandemFoil and `val_primary/surface_mse` for AirfRANS.

## Baseline (UPDATED after your PR #2899 merge)

- **TandemFoil val_primary/surface_pressure_mae:** **26.134** (PR #2899, ep123, EMA decay=0.999)
- **AirfRANS val_primary/surface_mse:** **0.000727** (PR #2899, ep206, EMA decay=0.999)
- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, no EMA)
- **Reproduce TF baseline:** `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT robin:

## Results

### Summary

| Run | Dataset | Config | Best val_primary | Best Epoch | Baseline | Δ% |
|-----|---------|--------|-----------------|------------|----------|----|
| 1 | TandemFoil | lr=1e-4, gc=1.0, EMA 0.999 | **24.526** | ep167 | 26.134 | **-6.2%** |
| 2 | TandemFoil | lr=1.25e-4, gc=0.5, EMA 0.999 | **22.537** | ep336 | 26.134 | **-13.8%** ✨ |
| 3 | AirfRANS | T_max=10, EMA 0.999 | **0.000659** | ep311 | 0.000727 | **-9.4%** ✨ |
| 4 | AirfRANS | T_max=5, EMA 0.999, seed=43 | **0.001090** | ep731 | 0.000727 | +49.9% |

### Run 1 — TandemFoil lr=1e-4 + EMA 0.999
- **Best val_primary/surface_pressure_mae:** 24.526 @ ep167 (-6.2% vs baseline 26.134)
- **W&B:** `rdsd8utw` — robin/ema-refine-tf-lr1e4
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent robin --wandb-name "robin/ema-refine-tf-lr1e4" --wandb-group robin-ema-refinement`
- **Final epoch:** 340 (360 min timeout)
- **Status:** ⚠️ Diverged after ep167. Lower LR + T_max=10 cosine cycling caused increasingly violent oscillations. By ep220+ val was spiking to 60–350. Grad norm reached inf at the end. The best at ep167 was captured during the first few stable cosine troughs, but the model was unable to maintain convergence at this LR.

### Run 2 — TandemFoil lr=1.25e-4 + gc=0.5 + EMA 0.999 ✨ NEW BEST
- **Best val_primary/surface_pressure_mae:** 22.537 @ ep336 (-13.8% vs baseline 26.134)
- **Test val (final epoch EMA):** test_eq4/surface_pressure_mae = 24.253
- **W&B:** `0lv7fnun` — robin/ema-refine-tf-gc05
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent robin --wandb-name "robin/ema-refine-tf-gc05" --wandb-group robin-ema-refinement`
- **Final epoch:** 336 (360 min timeout)
- **Status:** ✅ Stable throughout. gc=0.5 prevented the wild oscillations seen in Run 1, enabling consistent descent across 336 epochs. Final grad_norm 0.257, loss 0.057. The model was still finding new bests at ep336, suggesting even longer training could push further.

### Run 3 — AirfRANS T_max=10 + EMA 0.999 ✨ NEW BEST
- **Best val_primary/surface_mse:** 0.000659 @ ep311 (-9.4% vs baseline 0.000727)
- **Test (final epoch EMA):** full_test/surface_mse = 0.001260
- **W&B:** `2p71zpxm` — robin/ema-refine-af-tmax10
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --agent robin --wandb-name "robin/ema-refine-af-tmax10" --wandb-group robin-ema-refinement`
- **Final epoch:** 763 (360 min timeout)
- **Status:** ✅ T_max=10 with EMA 0.999 was a strong combination. Reached the first trough best (0.000659) at ep311 and maintained stability through ep763. The wider cosine cycle (T_max=10 vs T_max=5) avoided the divergence seen with T_max=5 in PR #2899, while still allowing deep exploration.

### Run 4 — AirfRANS T_max=5 + EMA 0.999 + seed=43
- **Best val_primary/surface_mse:** 0.001090 @ ep731 (+49.9% vs baseline 0.000727)
- **Test (final epoch EMA):** full_test/surface_mse = 0.003541
- **W&B:** `1v2wtbs8` — robin/ema-refine-af-seed43
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --seed 43 --agent robin --wandb-name "robin/ema-refine-af-seed43" --wandb-group robin-ema-refinement`
- **Final epoch:** 761 (360 min timeout)
- **Status:** ❌ Negative result. seed=43 with T_max=5 did not reproduce the 0.000727 result from seed=42 (PR #2899). Best was 0.001090 — still a strong result vs the pre-EMA baseline (0.001236) but well above the seed=42 best. This confirms that the T_max=5 + EMA interaction is seed-sensitive: the original 0.000727 hit a lucky cosine trough alignment that doesn't generalize.

### What happened

**TandemFoil gc=0.5 is a major new best.** The softer gradient clip (0.5 vs 1.0) with EMA was the winning combination, delivering a 13.8% improvement over the PR #2899 baseline. The gc=0.5 constraint prevented the catastrophic oscillations that destroyed Run 1 (lr=1e-4 + gc=1.0), while still allowing the model to descend steadily over 336 epochs. The model was still improving at timeout, suggesting even longer runs could push further.

**AirfRANS T_max=10 delivers a reliable new best.** The 0.000659 result at ep311 is a genuine improvement over T_max=5's 0.000727. Unlike the T_max=5 result (which was seed-sensitive, as Run 4 confirms), T_max=10 provides a wider, more stable cosine basin that EMA can average over effectively. This is a more robust finding.

**lr=1e-4 is too low for TandemFoil with T_max=10.** While it found a decent trough (24.53 @ ep167), the model diverged afterward. The cosine cycling at this lower LR creates wider oscillations that EMA cannot smooth out, eventually leading to gradient explosion.

### Suggested follow-ups

1. **TandemFoil gc=0.5 with longer training** — Run 2 was still descending at ep336. A longer timeout (or early stopping at best) could push below 22.
2. **TandemFoil gc=0.5 + lr=1e-4** — Combining both the softer clip AND lower LR might avoid the divergence seen in Run 1 while getting the benefit of finer optimization.
3. **AirfRANS T_max=10 as new default** — T_max=10 + EMA 0.999 is more robust than T_max=5 (not seed-dependent). Consider making this the standard config.
4. **Best-checkpoint saving** — Run 1's divergence highlights that we lose access to the best model weights when training diverges. A best-checkpoint mechanism would preserve peak performance.
5. **gc=0.5 across all datasets** — The stabilizing effect of softer gradient clipping deserves cross-dataset testing, especially for DrivAerML which showed EMA instability in PR #2899.

**Comment by tcapelle** [comment]
ADVISOR: Merging for the TandemFoil win.

**Run 2 (TF gc=0.5 + EMA 0.999): 22.537 — new TF best (-13.8% vs 26.06 global baseline)**. The softer gradient clip finding is important: gc=0.5 enables stable EMA training across 336+ epochs where gc=1.0 diverges at cosine troughs. Model still descending at timeout — this is an underestimate of the ceiling.

**Run 3 (AirfRANS T_max=10 + EMA 0.999): 0.000659** — this was a win at the time of submission but has been superseded. PR #2951 (stark) merged 0.000482 using AdamW with T_max=50 (no EMA). The T_max=50 cosine schedule is the critical finding: longer periods let the model fully descend before cycling. T_max=10 with EMA cannot match that.

Good analysis on the seed sensitivity of T_max=5. That confirms T_max=5 was a lucky alignment, not a robust result.

Follow-up suggestions are well-reasoned. The gc=0.5 cross-dataset test and longer-training TF run are both on the roadmap.
